### ISSUE TITLE
include_role does not support tags

### ISSUE THREAD

**Issue by mwoodson** [ansible/ansible-modules-core#5077]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

include_role
##### ANSIBLE VERSION

```
ansible 2.2.0
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

Linux, Red Hat Enterprise 7.2
##### SUMMARY

Originally filed this issue here because I was told on IRC this was part of ansible because it was a action plugin.  The bot told me i was wrong after a few days, so now it's here.

https://github.com/ansible/ansible/issues/17761

Forgive my terminology.  I will do my best to describe that issue I am seeing.  I am using a pre-release version of Ansible 2.2 for this.

When using include_role module with tags, the include_role seems to ignore the tags and does not apply them when being called.

I am using include_role in my playbook in the task list.  I apply a tag on it.  When I call ansible-playbook -t tagname it does not run my include_role task.
##### STEPS TO REPRODUCE

```

---
- name: test playbook
  hosts: localhost

  tasks:
  - debug:
      msg: "hello from debug"
    tags:
    - test

  - include_role:
      name: role1
    tags:
    - test
```

OUTPUT

```
$ ansible-playbook -t test test.yml
PLAY [test playbook] ***********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "hello from debug"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   

```
##### EXPECTED RESULTS

I would expect it to run the role1, because it was called with the tags
##### ACTUAL RESULTS

role1 was never called

**Comment by ansibot** [comment]
@ansible, @abadger, @adq, @AlanCoding, @alikins, @amenonsen, @angstwad, @anoek, @ansible-jenkins, @ansibull, @b3rn, @bbaassssiiee, @bcoca, @bennojoy, @carsongee, @cchurch, @chouseknecht, @chrishoffman, @chrismeyersfsu, @chrrrles, @criptoverso, @danvaida, @defionscode, @dfederlein, @dharmabumstead, @docschick, @elventear, @emonty, @eoinverling, @erjohnso, @fujiwarat, @gamuniz, @ghjm, @gregdek, @gundalow, @headmin, @j00bar, @j2sol, @jaimegago, @jainnikhil30, @jakemcdermott, @jamesmarshall24, @jangsutsr, @jaredevantabor, @jarv, @jbrockett, @jctanner, @jeremyollis, @jhawkesworth, @jimi-c, @jladdjr, @jlaska, @jlmitch5, @joshuaconner, @jsdalton, @juhp, @justnems, @jvanpelt, @kensible, @leigh-johnson, @lorin, @lwade, @lwojciknc, @mabashian, @manicminer, @matburt, @mattclay, @matts-mpg, @mckerrj, @mholiv, @michelleperz, @mscherer, @neptunerx, @nitzmahone, @petemounce, @phips, @privateip, @Qalthos, @resmo, @rmfitzpatrick, @robynbergeron, @ryansb, @s-hertel, @samdoran, @shanemcd, @shepdelacreme, @Shrews, @simfarm, @sivel, @skottler, @smadam813, @smapjb, @smashwilson, @softzilla, @srvg, @stphnwallace, @sundeep-co-in, @tbarr, @thaumos, @thisdavejohnson, @tima, @tjcramer, @tonk, @trahman73, @trondhindenes, @tylerturk, @wenottingham, @willthames, @wtcross, @wwitzel3, @zbal, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gregdek** [comment]
I'm gonna go out on a limb here and say that it's bad when ansibot is removing the bot_broken label. Is it self-aware?

**Comment by gregdek** [comment]
Hey bot: bot_broken

**Comment by ansibot** [comment]
@ansible ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mwoodson: Thanks for taking the time to open this issue, but it was filed under the wrong github repository. Please create a new issue at https://github.com/ansible/ansible-modules-ansible/issues
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sosiouxme** [comment]
ansible-modules-ansible doesn't seem to exist (anymore).

This seems to have been somewhat fixed in 2.3. I'm experiencing a possibly-related problem where if include_role has a tag and you pass in that on --tags, all of the tasks for that role and its dependencies run even if they aren't tagged. So, might still need some tweaking. Will open a separate issue...

**Comment by sosiouxme** [comment]
Strike that, it works [the same as other includes do](http://docs.ansible.com/ansible/playbooks_tags.html#tag-inheritance). It's a little surprising but the `include_role` task isn't itself selected/skipped by the tags it has on it; instead, any tag you specify on it is applied (recursively) to everything it includes, so everything it includes will have the tag and be selected by it.

**Comment by ocervell** [comment]
I'm running into similar issues with this `include_role` as well.

@sosiouxme It doesn't seem to work exactly the same as other includes do, regarding to tags.

**STEPS TO REPRODUCE**

***roles/common/tasks/main.yml***
```
- name: Task 1
  debug: name="task no.1"
  tags:
    - task_1

- name: Task 2
  debug: name="task no.2"
  tags:
    - task_2

- name: Task 3
  debug: name="task no.3"
  tags:
    - task_3
```

***play.yml***
```
- name: test playbook
  hosts: localhost
  tasks:
    - include_role:
         name: common
```

**EXPECTED RESULT**
The `common` role should executing "Task 1" when running:

`ansible-playbook play.yml --tags task_1`

**ACTUAL RESULTS**
The `include_role` statement is being skipped.

**TEMPORARY FIXES**
I was able to make this work was by explicitly listing all the tags used by the tasks within my role  in the `include_role` statement, like so:

```
- name: test playbook
  hosts: localhost
  tasks:
    - include_role:
         name: common
      tags:
         - task_1
         - task_2
         - task_3
```

... but this feels hacky because ***you have to go through every task in the role and list them*** ... This can be problematic when you have tons of tasks, each with different tags.

IMHO, another thing that ***should*** work in this case is specifying 'always' tags and expect the respective tasks in the role to honour their own tags:

***play.yml***
```
- name: test playbook
  hosts: localhost
  tasks:
    - include_role:
         name: common
      tags: always
```
... but with this, **all the tasks within the role will run ignoring whatever tags are passed on the CLI.**

**Comment by dagwieers** [comment]
@ocervell You understand that this repository is no longer functional and everything has moved to https://github.com/ansible/ansible

Since this is more like a proposal as well, you my want to create your own proposal at https://github.com/ansible/proposals

In any case, don't expect any feedback in this repository though. Most people have left this building...

**Comment by ocervell** [comment]
@dagwieers No, I wasn't aware. It's not a proposal, it's more a bug report, because the logical behaviour for this is not respected. I will add an issue to [https://github.com/ansible/ansible](https://github.com/ansible/ansible).
